### ISSUE TITLE
Returned data format

### ISSUE THREAD

**Issue by simonmoulds** [Ryan-Riggs/RivRetrieve#11]
One decision we need to make is the format of the data we return. I'm uneasy about removing the metadata that some of the web services return (e.g. quality flags). One option would be to return the original data (with columns as returned by the underlying web service or download function), but with some additional RivRetrieve-specific columns that are consistent across all the functions (and may duplicate some of the original column). These could be prefixed with `rr_` to differentiate them from the original data. E.g. 

| rr_date | rr_variable | rr_value | ... |
| --- | --- | --- | --- |
| ... | ... | ... | ... |

Users are then able to see the original data as well as have a consistent set of columns that they can use when programming. An advanced option here would be to define an S3 class inheriting from `tibble` with some additional attributes to identify the original/new columns. The nice thing about this option is that we can develop methods for the object, so you could do something like: 

```
x <- australia(site="<site-id>", variable="discharge")
plot(x)
``` 
And a nice hydrograph would be displayed. You could also have methods to compute indices, e.g. `bfi(x)` etc. 

Perhaps the above features are for another day, but the original point - the need to decide on a consistent format for the data returned from each function - is one we should address soon. 

Let me know your thoughts.

**Comment by Ryan-Riggs** [comment]
Yeah this is something I'm also wondering about. I've also been leery of removing the quality flags and other attributes but since they aren't consistent, I went ahead and removed them. I think your suggestion on adding 'rr_' to the RivRetrieve specific columns makes sense but the issue is that some of the raw data are in completely different formats. For example, the Brazilian data are oriented something like the following: 

| Month  | Year | Day 1 (Q) | Day 2 (Q) | Day ... (Q) | Day 31 (Q) |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| 1  | 2020 | 10.5 | 15.6 | 17.0 | 10.5 |
| 2  | 2020 | 12.3 | 10.1 | 15.1 | 12.5 |


What about returning the data in the current RivRetrieve format and adding in additional columns where available? Alternatively, we could just return a list with two data frames, the first being the RivRetrieve consistent format and the second being the raw data from the agency. Do you have any thoughts on this?

**Comment by simonmoulds** [comment]
Yes, I know what you mean. I think the first option - returning data in current RR format and adding additional columns where available is probably the best one. Including the raw data as an attribute to the returned dataframe is another option, but deciding where to draw the line between raw and processed could be tricky (e.g. the data from Japan requires a lot of processing before it can be used, and it's unclear what would constitute the raw dataset). 

If we pursue this option I would still advocate for prefixing the RivRetrieve columns with "rr_". This makes it easy for users to understand what is the original data and what is from RR. It also means we can impose a convention on units (m3/s for discharge, m for stage, I would suggest) for any data prefixed with "rr_". We could include an option in the functions like `include_original=TRUE`, so that users had the option of not attaching the original data. 

In summary, I would suggest:
- Including any metadata from the original data with minimal processing (just enough to get it to fit with the current RR convention of providing a two-column dataframe with time/data
- Prefix the RR columns with "rr_" to distinguish them from the original data
- Add an option to the various countries to allow users not to include the original data in the returned dataframe.

Let me know what you think.

**Comment by Ryan-Riggs** [comment]
I agree with all of these suggestions. I think it's important to add the 'rr_' prefix and I agree that including a consistent unit format (discharge=m3/s and stage=m) is necessary. 

I just want to make sure I follow your suggestion for include_original=TRUE. Which of the following are you suggesting?

1. Add additional columns into the RivRetrieve output that represent that raw data (colnames=c(rr_date, rr_q, raw_date, raw_q,...)).
2. Only return the raw data in a dataframe format. 
3. Return a list (or something similar) of two dataframes (RivRetrieve dataframe and raw dataframe). 

I think I'm personally leaning toward option 2 or 3. I think option 3 makes a lot of sense but I worry that people unfamiliar with R may be confused by a list of dataframes. What are your thoughts?

**Comment by simonmoulds** [comment]
I was actually thinking (1), but (3) is also a good option. Instead of a list you can do this with attributes. It's made very easy with [new_tibble()](https://tibble.tidyverse.org/reference/new_tibble.html). In our case, you would do something like:
```{R}
australia <- function(...) { 
  <do stuff to get data>
  x = new_tibble(
    rr_data, # The processed data
    original_data = orig_data, # The original data
    class = "rr_tbl"
  )
  return(x)
}
```
Now the original data is included as an attribute to the returned dataframe, but the object has all the functionality of a conventional tibble/data.frame. You can retrieve it with `attr(x, "original_data")`, but it's also straightforward to write a getter method:
```{R}
original <- function(x, ...) { 
  UseMethod("original")
}

original.rr_tbl <- function(x, ...) { 
  return(attr(x, "original_data"))
}
```
Then you get the original data by doing `original(x)`. In this case I probably wouldn't give users an option, and always include it.

**Comment by Ryan-Riggs** [comment]
Okay, that makes sense. I think option 3 would be ideal because then we could provide the true raw data in its native format and not have to bother with writing additional scripts to reformat the raw data. However, if you disagree, then I'm happy to discuss further. 

I don't really know much about the new_tibble method but it seems like it would be a good approach to take.
